cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java

```diff
@@ -0,0 +1,163 @@
+    this.cacheRefreshIntervalSecs = cConf.getInt(Constants.Security.Authorization.CACHE_REFRESH_INTERVAL_SECS);
+    this.authPolicyCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS), TimeUnit.SECONDS)
+      .build(new CacheLoader<Principal, Set<Privilege>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Set<Privilege> load(Principal principal) throws Exception {
+          return getPrivileges(principal);
+        }
+      });
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0, cacheRefreshIntervalSecs, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting authorization enforcement service...");
+    if (!authorizationEnabled) {
+      if (cacheEnabled) {
+        LOG.warn("Authorization policy caching is enabled ({} is set to true), however, no privileges will be cached " +
+                   "because authorization is disabled ({} is set to false). ",
+                 Constants.Security.Authorization.CACHE_ENABLED, Constants.Security.Authorization.ENABLED);
+      }
+      return;
+    }
+    updatePrivilegesOfCurrentUser();
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (!(authorizationEnabled && cacheEnabled)) {
+      return;
+    }
+    LOG.trace("Running authorization enforcement service iteration...");
+    for (Principal principal : authPolicyCache.asMap().keySet()) {
+      updatePrivileges(principal);
+    }
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    if (!authorizationEnabled) {
+      return;
+    }
+    if (!cacheEnabled) {
+      authorizerInstantiator.get().enforce(entity, principal, action);
+      return;
+    }
+    Set<Privilege> privileges = authPolicyCache.get(principal);
+    if (!privileges.contains(new Privilege(entity, action))) {
+      throw new UnauthorizedException(principal, action, entity);
+    }
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Shutting down authorization enforcement service...");
+    authPolicyCache.invalidateAll();
+  }
+
+  @VisibleForTesting
+  Map<Principal, Set<Privilege>> getCache() {
+    return authPolicyCache.asMap();
+  }
+
+  private Set<Privilege> getPrivileges(Principal principal) throws Exception {
+    return authorizerInstantiator.get().listPrivileges(principal);
+  }
+
+  /**
+   * On an authorization-enabled cluster, if caching is enabled too, updates the cache in the
+   * {@link AuthorizationEnforcementService} with the privileges of the user running the program.
+   */
+  private void updatePrivilegesOfCurrentUser() {
+    String userName;
+    try {
+      userName = UserGroupInformation.getCurrentUser().getShortUserName();
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```

Comment:
My thinking was:
1. This is done before the program starts running. Say due to a transient error, this operation fails.
2. When the program starts, and again requests for privileges from the cache, the LoadingCache does not have anything, so it makes a call to the authorization backend, and if the transient error is fixed before this first access from the program occurs, the program could still run fine. If it doesn't get fixed, then the user gets an error during program runtime, if he tries to execute an operation that requires authorization.

However, for this case, the scant javadoc for ``getCurrentUser()`` says that it throws ``IOException`` if login fails. That didn't seem like a transient error.

Do you think that overcomplicates it, and we should just standardize on failing the program if privileges can't be updated before it starts? If the above approach is ok, I can add a comment to explain that.